exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/aggregate/HashAggTemplate.java

```diff
@@ -1317,7 +1364,7 @@ private void checkGroupAndAggrValues(int incomingRowIdx) {
 
         useReservedValuesMemory(); // try to preempt an OOM by using the reserve
 
-        addBatchHolder(currentPartition);  // allocate a new (internal) values batch
+        addBatchHolder(currentPartition, getBatchSize());  // allocate a new (internal) values batch
```

Comment:
I don't really agree with this approach. We are dynamically adjusting the BatchHolder size based on new records we see coming in, but does this even make sense to do?

We don't really have control over which BatchHolder incoming records get hashed into, so they may get hashed into a BatchHolder that was size based on different input data. Therefore we don't actually get any tangible benefit for independently sizing batch holders dynamically, because the data they contain gets distributed randomly based on a hash. If there was a close correlation between the data in an input batch and the data in a BatchHolder, there would actually be a benefit, but that is not the case here.

I guess this was done because with the current code the size of an output batch is equivalent to the size of a batch holder since an entire batch holder is emitted at once. But instead of trying to work within that limitation I think we should actually change to code to allow the output of a partial batch holder. Additionally we should not allow heterogenous batch holder sizes. We could use the statistics from the first input batch to do a reasonable sizing for batch holders instead of the current 64k default, which would be a great change. However, after a BatchHolder size is selected it should not be changed.

Reply:
Adjusting batch holder size here means adjusting number of rows in the batch, based on average row width. Idea is to limit size in terms of memory, not in terms of number of rows. Batches are limited to 16MB (or whatever configured output batch size).  By allocating huge batches and partially transmitting them, we might be able to limit output batch size, but that does not produce much benefit. We want to avoid huge memory allocations.
Why we should not change batch holder size ? 
If we size just based on first batch, it creates the exact problem you mentioned i.e. they will be sized based on older input data and they may not make much sense for new data.
What I have is not exact perfect solution. In fact, I don't even know if such a solution is possible or exists. This will work fine with law of averages.

Reply:
Padma I agree we want to limit the size of output batches, and that reducing the batch holder size is a great change. Having BatchHolders with 64K always 64k rows is not practical. My issue is with changing the batch holder size dynamically. I think it adds complexity without a concrete benefit. Since new data will be added to old BatchHolders data will never really go into a BatchHolder that was appropriately sized for it. Since we can't really have an accurate solution by taking the complex approach with dynamically changing BatchHolder sizes, I think we should go with the simpler approach. We can still use all your changes, I just think we shouldn't continue updating the BatchHolder size.

The complexity added is in the added overhead for computing indexes in the hashtable, and there will be more complexity in doing the refactored memory calculations I am adding.

Reply:
agree. there is complexity with variable size batch holders. but, I already dealt with it and made the changes. idea is whenever we are allocating a new batch, we want to use the most recent information. Let us discuss offline with @Ben-Zvi also and decide what we want to do here.

Reply:
Yeah let's talk about it offline.

Reply:
Again confusion between **size** and **count** -- The new BatchHolder is allocated based on a **count** of rows, this count is computed (based on sizing the incoming) such that the result batch size would not exceed the given (e.g. 16MB). 
   Now worst case, the following incoming (or a size deviation within the averaged rows) yields bigger values (e.g. longer varchars) -- Looks like the current code would continue to fill the previous batch based on the original **count**. Maybe a check can be added so in this case fewer rows would be inserted into the batch (i.e., less than the original count), such that the **size** restriction would be kept.
   Hopefully such a check would be cheap ....